File: backend/checker/voice.py
"""
How the tutor talks.

The maths is the same whoever is being taught — it is checked by SymPy and does
not have a mood. What changes is the wrapping around it, and that matters more
than it sounds. A student who has been stuck for twenty minutes and a student
who wants the answer and nothing else are both badly served by the same voice,
and the wrong one reads as either patronising or cold.

So this is only ever added to the end of an instruction. It can change the
warmth, the length, and how much is spelled out. It cannot change what counts
as a right answer, and nothing here is allowed to loosen the rule that the
tutor never gives away the answer to the question in front of them.
"""

ENCOURAGING = """
How to say it: warmly. This student loses heart quickly, so when you have to
name a mistake, notice what they got right first, and mean it rather than
reaching for praise you do not have. Being wrong is the ordinary way of
learning something, and your tone should take that for granted. Never sound
disappointed.

Warmth is a tone and not an opening line. If they have asked you something,
answer it in your first sentence. Praise standing in front of an answer pushes
the answer off the end of the page, and reads as though you did not listen.
""".strip()

DIRECT = """
How to say it: briefly. This student wants the point and nothing else. No
preamble, no reassurance, no "nice try". One sentence is usually enough. Being
brief is not the same as being cold: say it plainly, the way you would to
someone you respect who is in a hurry.
""".strip()

THOROUGH = """
How to say it: fully. This student wants to understand why, not just what. Give
the reason behind the rule as well as the rule. When you are naming a mistake,
say what would have gone wrong if they had carried on. Take an extra sentence
where an extra sentence earns its place, but do not pad, and do not explain
what they have already shown they know.
""".strip()

PUNCTUATION = """
Write in plain sentences. No dashes standing in for punctuation, no semicolons,
no bullet points: end the sentence and start another one. Every word of this is
copied out in handwriting on a page of ruled paper by a tutor sitting next to
the student, and none of that punctuation is something a person writes by hand.
""".strip()

VOICES = {
    "encouraging": ENCOURAGING,
    "direct": DIRECT,
    "thorough": THOROUGH,
}

DEFAULT = "encouraging"


def spoken(instructions, style):
    """
    The instruction, with how to say it added on the end.

    An unknown style is not an error. It arrives from the app, and an app one
    version ahead of the server asking for a voice this server has never heard
    of should still get taught — in the ordinary voice, not a 500.
    """
    voice = VOICES.get((style or DEFAULT).strip().lower())

    if voice is None:
        return f"{instructions}\n\n{PUNCTUATION}"

    return f"{instructions}\n\n{PUNCTUATION}\n\n{voice}"

File: StepOut/StepOut/Voice.swift
import Foundation

/// How the student wants the tutor to talk to them.
///
/// The marking is the same either way — it is done by a computer algebra
/// system and does not have a mood. This changes only the wrapping around it,
/// which is not nothing: a student who has been stuck for twenty minutes and
/// one who wants the mistake named and nothing else are badly served by the
/// same voice, and the wrong one reads as either patronising or cold.
enum Voice: String, CaseIterable, Identifiable {
    case encouraging
    case direct
    case thorough

    var id: String { rawValue }

    var name: String {
        switch self {
        case .encouraging: "Encouraging"
        case .direct: "Straight to the point"
        case .thorough: "Explain everything"
        }
    }

    var summary: String {
        switch self {
        case .encouraging:
            "Notices what you got right before naming what went wrong."
        case .direct:
            "Names the mistake in a sentence. No warm-up."
        case .thorough:
            "Gives the reason behind the rule, not just the rule."
        }
    }

    var symbol: String {
        switch self {
        case .encouraging: "hand.thumbsup"
        case .direct: "bolt"
        case .thorough: "book"
        }
    }
}
